[SPARK-8015] [flume] Remove Guava dependency from flume-sink.

vanzin opened the pull request:
The minimal change would be to disable shading of Guava in the module,
and rely on the transitive dependency from other libraries instead. But
since Guava's use is so localized, I think it's better to just not use
it instead, so I replaced that code and removed all traces of Guava from
the module's build.

Commit:
[SPARK-8015] [flume] Remove Guava dependency from flume-sink.

The minimal change would be to disable shading of Guava in the module,
and rely on the transitive dependency from other libraries instead. But
since Guava's use is so localized, I think it's better to just not use
it instead, so I replaced that code and removed all traces of Guava from
the module's build.

vanzin commented:
I verified that there are no references to Guava in the generated bytecode and that Guava does not show up in `mvn dependency:tree`.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33905 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33905/consoleFull) for   PR 6555 at commit [`2d79260`](https://github.com/apache/spark/commit/2d79260269543a08074bdaa34e8486151061f722).

SparkQA commented:
[Test build #33905 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33905/consoleFull) for   PR 6555 at commit [`2d79260`](https://github.com/apache/spark/commit/2d79260269543a08074bdaa34e8486151061f722).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

squito commented on external/flume-sink/pom.xml:
```diff
@@ -85,6 +97,13 @@
+        <artifactId>maven-shade-plugin</artifactId>
+        <configuration>
+          <relocations combine.self="override" />
+        </configuration>
```
is it worth adding a short comment here explaining why this is necessary?  maybe this is obvious to others more versed in maven, but it was a little mysterious to me.  Then again we don't doc a lot of the intricacies of the build system.  anyway, just a suggestion, I'll leave it up to your judgement.

squito commented:
jenkins, retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33907 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33907/consoleFull) for   PR 6555 at commit [`2d79260`](https://github.com/apache/spark/commit/2d79260269543a08074bdaa34e8486151061f722).

SparkQA commented:
[Test build #33907 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33907/consoleFull) for   PR 6555 at commit [`2d79260`](https://github.com/apache/spark/commit/2d79260269543a08074bdaa34e8486151061f722).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

Commit:
Add comment in pom.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33908 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33908/consoleFull) for   PR 6555 at commit [`6e0942d`](https://github.com/apache/spark/commit/6e0942d954fcaebe34c91d183d23485aea79ed92).

harishreedharan commented:
+1. LGTM

SparkQA commented:
[Test build #33908 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33908/consoleFull) for   PR 6555 at commit [`6e0942d`](https://github.com/apache/spark/commit/6e0942d954fcaebe34c91d183d23485aea79ed92).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

vanzin commented:
Hmm... the build failure doesn't seem related to the changes, but I'll take another look.

Commit:
Add libthrift exclusion.

Not sure what's wrong, but running sbt with no profiles enabled causes
an error about not being able to resolve version "${thrift.version}" of
libthrift.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33909 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33909/consoleFull) for   PR 6555 at commit [`b7a0349`](https://github.com/apache/spark/commit/b7a034973d9e3b215272a9937a2ba9ea8d2eee2f).

SparkQA commented:
[Test build #33909 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33909/consoleFull) for   PR 6555 at commit [`b7a0349`](https://github.com/apache/spark/commit/b7a034973d9e3b215272a9937a2ba9ea8d2eee2f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

harishreedharan commented:
Some of the test code's transitive dependencies seem to pull in `Preconditions`. How about including Guava as a test dependency? Does that complicate things?

vanzin commented:
Hmm. Wonder how those tests passed locally. Yeah, I'll add the dependency in test scope.

Commit:
Add guava dep in test scope.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33914 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33914/consoleFull) for   PR 6555 at commit [`c38228d`](https://github.com/apache/spark/commit/c38228d22266bef1680373f432007ea2419e4c4b).

SparkQA commented:
[Test build #33914 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33914/consoleFull) for   PR 6555 at commit [`c38228d`](https://github.com/apache/spark/commit/c38228d22266bef1680373f432007ea2419e4c4b).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

tdas commented on external/flume-sink/pom.xml:
```diff
@@ -42,22 +42,47 @@
-        Flume 1.4.0 depends on is "3.4.0.Final" .
-      -->
+      <groupId>com.google.guava</groupId>
+      <artifactId>guava</artifactId>
```
Why do you need to add this in test?

tdas replied:
I see, I missed the earlier conversation. But instead of randomly adding guava back for test, its better to remove the wrong version of guava that is coming in from somewhere, and allow Flume's guava dependency to be used. At least, it needs to be understood. Could you check the dependency tree to find the wrong guava version?

vanzin replied:
There's no wrong version of Guava.

I added this because I removed Guava from compile-scope dependencies, to avoid having people use it in this module. Not because of a conflict.

If instead you'd rather keep it, but at the same version as flume (whatever that is), it could be done also. I kinda prefer avoiding it, though.

As I mention in the PR description, the minimal fix doesn't remove Guava at all, just removes the code to relocate it. I just thought it would be better to avoid another potential dependency mess here since it's not needed.

tdas replied:
Aah, I was misunderstanding the problem. It makes sense now. Can you add comments on why Guava and thrift has been excluded.

Commit:
Add comments about dependency management.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33994 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33994/consoleFull) for   PR 6555 at commit [`c0ceea8`](https://github.com/apache/spark/commit/c0ceea81709e9ec8eb4229448fa5d473c4064963).

srowen commented:
Lgtm pending tests

SparkQA commented:
[Test build #33994 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33994/consoleFull) for   PR 6555 at commit [`c0ceea8`](https://github.com/apache/spark/commit/c0ceea81709e9ec8eb4229448fa5d473c4064963).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

tdas commented:
LGTM. Merging this, in master and branch 1.4.
